### PR TITLE
Fix alignment exceptions issue for F and C extensions

### PR DESCRIPTION
**Pull request by visitorckw**
This pull request address two issues related to data/instruction address alignment exceptions. The first patch adds address alignment checks for F extension load/store operations, resolving an issue reported by @fourcolor. The second patch fixes incorrect alignment checks in the C extension, aligning behavior with [RISC-V specifications](https://riscv.org/wp-content/uploads/2017/05/riscv-spec-v2.2.pdf) by eliminating unnecessary misalignment checks for jump instructions.

### CODE REVIEW

File: src/emulate.c [hunk 1]
```diff
@@ -38,13 +38,14 @@ extern struct target_ops gdbstub_ops;
...
 /* RISC-V exception code list */
-#define RV_EXCEPTION_LIST                                       \
-    _(insn_misaligned, 0)  /* Instruction address misaligned */ \
-    _(illegal_insn, 2)     /* Illegal instruction */            \
-    _(breakpoint, 3)       /* Breakpoint */                     \
-    _(load_misaligned, 4)  /* Load address misaligned */        \
-    _(store_misaligned, 6) /* Store/AMO address misaligned */   \
-    _(ecall_M, 11)         /* Environment call from M-mode */
+#define RV_EXCEPTION_LIST                                           \
+    IIF(RV32_HAS(EXT_C))                                            \
```
Comment (qwe661234) [inline src/emulate.c]: Why we need to remove all exception handlers under C extension? All exception handlers are still needed for other instructions.

File: src/rv32_template.c [hunk 2]
```diff
@@ -2158,7 +2163,6 @@ RVOP(
     clwsp,
```
Comment (qwe661234) [inline src/rv32_template.c]: The spec only mentions jump-related instruction, but clwsp and cswsp are load/store instruction.

File: src/emulate.c (hunk 1 above)
Comment (jserv) [inline src/emulate.c:42]: Use `/* clang-format off */` and `/* clang-format on */` to enforce consistent style.

Comment (jserv) [inline src/emulate.c]: Maybe we should consult Clang-Format style options for better indention. See https://clang.llvm.org/docs/ClangFormatStyleOptions.html

File: src/rv32_template.c [hunk 3]
```diff
@@ -238,51 +242,54 @@ RVOP(
...
-                goto end_insn;                                     \
-        }, );                                                      \
-        last_pc = PC;                                              \
-        MUST_TAIL return taken->impl(rv, taken, cycle, PC);        \
-    }                                                              \
-    end_insn:                                                      \
-    rv->csr_cycle = cycle;                                         \
-    rv->PC = PC;                                                   \
+#define BRANCH_FUNC(type, cond)                                              \
+    IIF(RV32_HAS(EXT_C))                                                     \
```
Comment (jserv) [inline src/rv32_template.c]: Consider the following changes:
```diff
--- a/src/rv32_template.c
+++ b/src/rv32_template.c
@@ -243,10 +243,8 @@ RVOP(
 /* clang-format on */
 
 #define BRANCH_FUNC(type, cond)                                              \
-    IIF(RV32_HAS(EXT_C))                                                     \
-    (, const uint32_t pc =                                                   \
-           PC;) if (BRANCH_COND(type, rv->X[ir->rs1], rv->X[ir->rs2], cond)) \
-    {                                                                        \
+    IIF(RV32_HAS(EXT_C))(, const uint32_t pc = PC;);                         \
+    if (BRANCH_COND(type, rv->X[ir->rs1], rv->X[ir->rs2], cond)) {           \
         is_branch_taken = false;                                             \
         struct rv_insn *untaken = ir->branch_untaken;                        \
         if (!untaken)  
```

File: src/emulate.c (hunk 1 above)
Comment (jserv) [reply src/emulate.c]: Alternatively, place `/* clang-format off */`.

**Review by qwe661234 — APPROVED** [review]

File: src/rv32_template.c (hunk 2 above)
Comment (visitorckw) [reply src/rv32_template.c]: Not really sure why I didn't see these comments until now.
Yes, it was my mistake, but it should be fixed now.
Thank you for your review.

File: src/emulate.c [hunk 4]
```diff
@@ -38,13 +38,17 @@ extern struct target_ops gdbstub_ops;
...
 
 /* RISC-V exception code list */
-#define RV_EXCEPTION_LIST                                       \
-    _(insn_misaligned, 0)  /* Instruction address misaligned */ \
-    _(illegal_insn, 2)     /* Illegal instruction */            \
-    _(breakpoint, 3)       /* Breakpoint */                     \
-    _(load_misaligned, 4)  /* Load address misaligned */        \
-    _(store_misaligned, 6) /* Store/AMO address misaligned */   \
+/* clang-format off */
+#define RV_EXCEPTION_LIST                                            \
```
Comment (jserv) [inline src/emulate.c]: Tweak the style below:
```diff
--- a/src/emulate.c
+++ b/src/emulate.c
@@ -39,10 +39,10 @@ extern struct target_ops gdbstub_ops;
 
 /* RISC-V exception code list */
 /* clang-format off */
-#define RV_EXCEPTION_LIST                                            \
-    IIF(RV32_HAS(EXT_C))(                                            \
-        , _(insn_misaligned, 0) /* Instruction address misaligned */ \
-    )                                                                \
+#define RV_EXCEPTION_LIST                                          \
+    IIF(RV32_HAS(EXT_C))(,                                         \
+        _(insn_misaligned, 0) /* Instruction address misaligned */ \
+    )                                                              \
```

File: src/emulate.c [hunk 5]
```diff
@@ -346,13 +350,7 @@ static block_t *block_find(const block_map_t *map, const uint32_t addr)
 
 FORCE_INLINE bool insn_is_misaligned(uint32_t pc)
```
Comment (jserv) [inline src/emulate.c:352]: I wonder if `insn_is_misaligned` is really used via macro expansion or not. Can you check?

Comment (visitorckw) [reply src/emulate.c:352]: I inserted a ```fprintf``` statement inside ```insn_is_misaligned()``` to verify if the function is being executed. It should not run during ```make ENABLE_EXT_C=1 check``` but should run during ```make ENABLE_EXT_C=0 check```. My experimental results align with my expectations.

Comment (jserv) [reply src/emulate.c:352]: Then, `insn_is_misaligned` should be conditionally built, depending on the configurations.

**Comment by jserv** [comment]
I tried executing the command `make ENABLE_EXT_C=0 ENABLE_JIT=1 clean all`, but it resulted in the following errors:
```
In file included from src/jit.c:1314:
src/rv32_jit.c:609:50: error: no member named 'shamt' in 'struct rv_insn'
emit_alu32_imm8(state, 0xc1, 5, temp_reg[0], ir->shamt);
                                             ~~  ^
src/jit.c:1312:9: note: expanded from macro 'GEN'
        code;                                                                 \
        ^~~~
In file included from src/jit.c:1314:
src/rv32_jit.c:614:50: error: no member named 'shamt' in 'struct rv_insn'
emit_alu32_imm8(state, 0xc1, 7, temp_reg[0], ir->shamt);
                                             ~~  ^
src/jit.c:1312:9: note: expanded from macro 'GEN'
        code;                                                                 \
        ^~~~
2 errors generated.
make: *** [build/jit.o] Error 1
```

@qwe661234, can you help confirm?

**Comment by visitorckw** [comment]
> I tried executing the command `make ENABLE_EXT_C=0 ENABLE_JIT=1 clean all`, but it resulted in the following errors:
> 
> ```
> In file included from src/jit.c:1314:
> src/rv32_jit.c:609:50: error: no member named 'shamt' in 'struct rv_insn'
> emit_alu32_imm8(state, 0xc1, 5, temp_reg[0], ir->shamt);
>                                              ~~  ^
> src/jit.c:1312:9: note: expanded from macro 'GEN'
>         code;                                                                 \
>         ^~~~
> In file included from src/jit.c:1314:
> src/rv32_jit.c:614:50: error: no member named 'shamt' in 'struct rv_insn'
> emit_alu32_imm8(state, 0xc1, 7, temp_reg[0], ir->shamt);
>                                              ~~  ^
> src/jit.c:1312:9: note: expanded from macro 'GEN'
>         code;                                                                 \
>         ^~~~
> 2 errors generated.
> make: *** [build/jit.o] Error 1
> ```
> 
> @qwe661234, can you help confirm?

The current master branch also encounters the same error.

**Comment by jserv** [comment]
> The current master branch also encounters the same error.

Let's create an issue which describes the above build failure.

**Comment by jserv** [comment]
Thank @visitorckw for contributing!

**Pull request merged**